### ISSUE TITLE
Spring Security and OIDC with Google - token retrieval fails

### ISSUE THREAD

**Issue by berndgoetz** [eugenp/tutorials#4037]
I'm trying to get the tutorial code running from the article http://www.baeldung.com/spring-security-openid-connect - but it fails with the POST call to the token endpoint.

The configuration is as follows:
`
server.port=8080
google.clientId=xxx
google.clientSecret=yyyy
google.accessTokenUri=https://www.googleapis.com/oauth2/v3/token
google.userAuthorizationUri=https://accounts.google.com/o/oauth2/auth
google.redirectUri=http://localhost:8080/google-login
google.issuer=accounts.google.com
google.jwkUrl=https://www.googleapis.com/oauth2/v2/certs
logging.level.org.springframework.security=DEBUG
`

The logs show:
`
2018-04-17 23:51:48.895 DEBUG 15312 --- [nio-8080-exec-3] g.c.AuthorizationCodeAccessTokenProvider : Retrieving token from https://www.googleapis.com/oauth2/v3/token
2018-04-17 23:54:07.223 DEBUG 15312 --- [nio-8080-exec-3] .a.SimpleUrlAuthenticationFailureHandler : No failure URL set, sending 401 Unauthorized error
`

The code in the RESTTemplate includes the three correct uriVariables, but the expanded URI does not contain those query parameters afterwards. That's why the call fails with a 401.
`
...
URI expanded = getUriTemplateHandler().expand(url, uriVariables);
return doExecute(expanded, method, requestCallback, responseExtractor);
`

The values of the uriVariables:
`"grant_type" -> "authorization_code"
"code" -> "4/AAAjBg1a_93EJfdGtwbn7afPRjYy4yHKzvTGtK9NMRLuAA0Sfqt2sXo1t1xmY-rMi85xzsLBGizNca39OVGqvTM"
"redirect_uri" -> "http://localhost:8080/google-login"`

The value of the expanded variable: https://www.googleapis.com/oauth2/v3/token

Something seems fishy with the RESTTemplate configuration. Any insights?

**Comment by saipraveengithub** [comment]
I am facing the same issue.

Followed the steps in the readme to setup the clientId and clientSecret for my app from the google developer console. However, there is a 401 error

This application has no explicit mapping for /error, so you are seeing this as a fallback.
There was an unexpected error (type=Unauthorized, status=401).
Authentication Failed: Could not obtain access token

The authorization code is sent as a param to the redirect url; it looks like the step to fetch the token from the code is failing. Not sure about the reason.
Please let me know what is the issue and how to resolve this

http://localhost:8080/login?state=lPUTMz&code=4/AAABGwcvimYggXTgS_PE_FA0Kj5yC51BMDbWwMtoHapeF3KV0ZlSQ65fMZymLE6KUkMyNp_4US1pn7eW2UKU5g0&authuser=0&session_state=916976b3d98f07bb514aee7cd69208b78ec0c762..b804&prompt=consent#

**Comment by lor6** [comment]
I'm not able to reproduce the issue, so this is quite strange.

Can you try with these credentials:
google.clientId=475873350264-g1opb20lf2fc60h0o84rrkn972krgkvo.apps.googleusercontent.com
google.clientSecret=GiA1Agf_aSt-bhTrnXjre-5Z

Just to check if it's an issue with the app configuration on the Console.

I'll look into this more. Thanks.

**Comment by berndgoetz** [comment]
Quick heads up: I think it doesn't work because I'm running my service behind a company firewall that blocks the access token url call. I didn't have time to validate it yet but it might be useful to mention it in the example/tutorial.

**Comment by eugenp** [comment]
That's an interesting note @berndgoetz  - but I'm not sure how much the tutorial can help with that. This tutorial (and any other out there) assumes that network traffic goes through. If we're talking about an internal environment where that's not the case - then that needs to chance, or you need to build your system around it - but I don't think the tutorial can help here. 

Hope that makes sense. 
Cheers, 
Eugen.

**Comment by berndgoetz** [comment]
> That's an interesting note @berndgoetz - but I'm not sure how much the tutorial can help with that. This tutorial (and any other out there) assumes that network traffic goes through. If we're talking about an internal environment where that's not the case - then that needs to chance, or you need to build your system around it - but I don't think the tutorial can help here.
> 
> Hope that makes sense.
> Cheers,
> Eugen.

I agree, the tutorial cannot help with that, but what would help is a note like "Be aware of the possible issue when you're sitting behind an Internet proxy, typically found in Enterprise environments" or similar.

Best regards,
Bernd

**Comment by eugenp** [comment]
Interesting point, @berndgoetz. 
My hesitation in adding that note is - that would be applicable to a lot of articles, right? What I mean is - most articles that cover any sort of communication over a network would need that note - since they might all be impacted by a strict firewall environment. 
So, the assumption of a standard network is, of course, a simplification, but it's a reasonable one in a tutorial that's not really focused on that. 

Does that make sense, or is there anything specific here I should be aware of?

**Comment by berndgoetz** [comment]
I agree with you, if all articles pointed that out, it would be quite redundant. But I believe with your articles, this is different, especially when it comes to an article about how to test drive the OIDC authorization code flow (where my local Spring Boot server makes a callout to the Internet, many of us would initially not be aware of). Or maybe it is worth a completely separate (though small) article in your series, titled like "How to do server to server calls through a (Enterprise) firewall / proxy configuration".

I generally see many developers behind company firewalls and filters suffer from issues that they can't just run tutorials out there in the Internet, so I think it just makes sense to also address them explicitely. I can't recall where I've read that statement, but most code written today is still custom code written behind company firewalls. Everything always works fine when I'm using my private laptop.

Long story short: I really love your tutorials, they are a great added value for me. You often appear high in my search results with very relevant content.

**Comment by eugenp** [comment]
Hey @berndgoetz - this might help: https://www.baeldung.com/java-connect-via-proxy-server

I'm glad you're enjoying the site - that's always good to know :)
